pages/chaoss/visualizations/contrib_importance_over_time.py

```diff
@@ -0,0 +1,436 @@
+    :param colname: A string representing the name of the column to screen patterns out of
+
+    returns a DataFrame with columns that matched, and column used to match, dropped.
+    '''
+    # filter for rows in df where the column value contains any of the specifed patterns
+    return df[~df[colname].isin(patterns)]
+
+def count_cntrbs(df, patterns, colname='login', ascending=False):
+    '''
+    :param df: A DataFrame 
+    :param contribution_type: A string indicating the type of contribution
+    :param patterns: A list representing of strings or regular expression, patterns to be screened out
+    :param colname: A string representing the name of column to screen patterns out of
+    :param descending: A boolean indicating the order in which the returned df columns are sorted according to contributor importance, 
+                      by default it is set to False which corresponds to descending order
+
+    returns a DataFrame of cntrb_id's and a count of each of their contributions, sorted
+    '''
+    # remove potential bots
+    df = filter_rows(df, patterns, colname)
+    # count the number of contributions per contributor 
+    df = (df.groupby('cntrb_id')["Action"].count()).to_frame()
+    # sort rows according to amount of contributions
+    df.sort_values(by='cntrb_id', ascending=True, inplace=True)
+    df = df.reset_index()
+    return df
+
+def calc_cntrb_importance(df, threshold):
+    '''
+    :param df: A DataFrame of cntrb_id's and contributions, sorted in descending order of contributions
+    :param contribution_type: A String indicating the activity of interest
+    :param threshold: A double indicating the cut off threshold for contributor importance
+
+    A contributor is ranked by how large of a proportion of all work they did - 
+    if someone authored 350 of 1000 commits, for instance, they'll account for 35% 
+    of the commits made. The list of percent contributions is ranked in descending order. 
+    The ordered list is greedily processed such that the smallest list of contributors is 
+    found that accounts for threshold percent of the work done for a given contribution type.
+   
+    returns the contributor importance, cntrb_id's and number of outliers for the given contribution type
+    '''
+    # identify and count outliers which we define as 2 standard deviations or more above from the mean
+    num_outliers = len(df[(stats.zscore(df["Action"]) > 2)])
+    
+    # calculate total number of contributions
+    t_cntrbs = df["Action"].sum()
+
+    # initilize empty list of to store contibutor ids for those who make up the contributor importance
+    cntrb_ids = []
+    # initialize cumulative percent of contributions
+    cum_per = 0
+
+    # add column name cntrbs to calculate the ratio of activity to the total amount of said activity
+    df["cntrbs"] = df["Action"] / t_cntrbs
+    df = df.reset_index()
+    
+    # iterate through each row in df and create a running sum of percent contributions
```

Comment:
I might be mis interpreting this but I dont know if this is actually getting bus factor. From reading this is seems like it is getting a possible number of contributors that meets the threashold but I dont see how it for sure gets the minimum number.

Reply:
Either way I feel like this can be greatly simplified as well by df put into this function, summing by cntrb_id, and ordering from greatest to least. Then multiply t_cntrbs by the threshold (lets call this cntrb_thresh) and sum the column with the contribution counts until it is greater than or equal to cntrb_thresh